webrtc/stun/stun_connection.nim

```diff
@@ -0,0 +1,205 @@
+  # Check for error from a BindingRequest message.
+  # Returns an option with some BindingErrorResponse if there is an error.
+  # Returns none otherwise.
+  # https://datatracker.ietf.org/doc/html/rfc5389#section-10.1.2
+  var res = StunMessage(msgType: StunBindingErrorResponse,
+                        transactionId: msg.transactionId)
+  if msg.getAttribute(AttrMessageIntegrity).isNone() or
+     msg.getAttribute(AttrUsername).isNone():
+    res.attributes.add(ErrorCode.encode(ECBadRequest))
+    return some(res)
+
+  # This check is related to the libp2p specification.
+  # Might be interesting to add a customizable function check.
+  let username = string.fromBytes(msg.getAttribute(AttrUsername).get().value)
+  let usersplit = username.split(":")
+  if usersplit.len() != 2 and not usersplit[0].startsWith("libp2p+webrtc+v1/"):
+    res.attributes.add(ErrorCode.encode(ECUnauthorized))
+    return some(res)
+
+  # https://datatracker.ietf.org/doc/html/rfc5389#section-15.9
+  var unknownAttr: seq[uint16]
+  for attr in msg.attributes:
+    let typ = attr.attributeType
+    if typ.isRequired() and typ notin StunAttributeEnum:
+      unknownAttr.add(typ)
+  if unknownAttr.len() > 0:
+    res.attributes.add(ErrorCode.encode(ECUnknownAttribute))
+    res.attributes.add(UnknownAttribute.encode(unknownAttr))
+    return some(res)
+
+  return none(StunMessage)
+
+# - Stun Messages Handler -
+
+proc stunMessageHandler(self: StunConn) {.async: (raises: [CancelledError]).} =
+  # Read indefinitely Stun messages from stunMsgs queue.
+  # Sends a BindingResponse or BindingResponseError after receiving a BindingRequest.
+  while true:
+    let message = await self.stunMsgs.popFirst()
+    try:
+      let decoded = StunMessage.decode(await self.stunMsgs.popFirst())
+      if decoded.msgType == StunBindingErrorResponse:
+        trace "Received a STUN error", decoded, remote = self.raddr
+        continue
+      elif decoded.msgType == StunBindingResponse:
+        # TODO: Process StunBindingResponse doesn't seem necessary for libp2p-webrtc-direct.
+        # Some browsers could be uncooperative. In that case, it should be implemented.
+        # It should be implemented for libp2p-webrtc.
+        continue
+      else:
+        let errorOpt = checkForError(decoded)
+        if errorOpt.isSome():
+          let error = errorOpt.get()
+          await self.conn.write(self.raddr, error.encode())
+          continue
+
+        let bindingResponse = self.getBindingResponse(decoded)
+        await self.conn.write(
+          self.raddr,
+          bindingResponse.encode(decoded.getAttribute(AttrUsername))
+        )
+    except SerializationError as exc:
+      warn "Failed to decode the Stun message", error=exc.msg, message
+    except WebRtcError as exc:
+      trace "Failed to write the Stun response", error=exc.msg
+
+proc init*(
+    T: type StunConn,
+    conn: UdpConn,
+    raddr: TransportAddress,
+    iceControlling: bool,
```

Comment:
Could you please add a comment explaining this property? I see in the tests it's used sometimes with `true` or `false`, but I don't understand what's the difference. It is also good to name the parameters when calling this proc.

Reply:
Done.